src/main/scala-2.11/markets/participants/OrderCanceler.scala

```diff
@@ -15,48 +15,37 @@ limitations under the License.
 */
 package markets.participants
 
-import akka.actor.Scheduler
-
-import scala.concurrent.duration.FiniteDuration
-import scala.concurrent.ExecutionContext
+import markets.{Canceled, Cancel}
+import markets.orders.Order
 
 
 /** Mixin Trait providing behavior necessary to cancel outstanding orders. */
 trait OrderCanceler extends MarketParticipant {
 
-  def submitOrderCancellation(): Unit
-
-  /** Schedule order cancellation.
-    *
-    * @param scheduler
-    * @param initialDelay
-    * @param executionContext
-    */
-  def scheduleOrderCancellation(scheduler: Scheduler,
-                                initialDelay: FiniteDuration)
-                               (implicit executionContext: ExecutionContext): Unit = {
-    scheduler.scheduleOnce(initialDelay, self, SubmitOrderCancellation)(executionContext)
-  }
-
-  /** Schedule order cancellation
-    *
-    * @param scheduler
-    * @param initialDelay
-    * @param interval
-    * @param executionContext
-    */
-  def scheduleOrderCancellation(scheduler: Scheduler,
-                                initialDelay: FiniteDuration,
-                                interval: FiniteDuration)
-                               (implicit executionContext: ExecutionContext): Unit = {
-    scheduler.schedule(initialDelay, interval, self, SubmitOrderCancellation)(executionContext)
-  }
+  def orderCancellationStrategy(): Option[Order]
 
   override def receive: Receive = {
-    case SubmitOrderCancellation => submitOrderCancellation()
+    case Canceled(order, _, _) =>
+      outstandingOrders -= order
+    case SubmitOrderCancellation =>
+      orderCancellationStrategy() match {
+        case Some(order) =>
+          val orderCancellation = cancel(order)
+          submit(orderCancellation)
+        case None =>  // no outstanding orders to cancel!
+      }
     case message => super.receive(message)
   }
 
-  private object SubmitOrderCancellation
+  /** Submit an order cancellation to a market. */
+  private def submit(cancellation: Cancel): Unit = {
+    val market = markets(cancellation.order.tradable)
+    market tell(cancellation, self)
+  }
+
+  /** Generate an order cancellation message. */
+  private def cancel(order: Order): Cancel = {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [cancel should have a smaller scope by using private[this]](https://www.codacy.com/app/davidrpugh/markets-sandbox/file/1704906047/issues/source?bid=2921361&fileBranchId=3046267#l47)